tests/test_local_executor.py

```diff
@@ -360,8 +382,10 @@ def test_finalize_failed_137(docker_cleanup, job, tmp_work_dir, volume_api):
     # impersonate an admin
     docker.kill(local.container_name(job))
 
+    # slightly strange that this is EXECUTED and not ERROR
     wait_for_state(api, job, ExecutorState.EXECUTED)
```

Comment:
ExecutoreState.ERROR means something went wrong with the **executor**, not with the job. From the executor's POV, the job has finished with an exit code, just like normal.

Reply:
but compare with test_running_job_terminated_finalized - a running job that is terminated via `terminate` puts us into ExecutorState.ERROR *not* EXECUTED. https://github.com/opensafely-core/job-runner/pull/588/files#diff-c0a1e0861f681ae3cbc247887b135901340f6501175217aa4903ef7f7af1068bR473 
Why are these two scenarios different?

Reply:
Hmm. 

One is the result of docker kill, which simulate a process ending asynchronously external to the API. I.e. we'll only be notified of it next time we call `get_status()`.

The other is the result of an explicit api.terminate() call, which blocks on termination, and we know is done. 

However, I think the current implementation of terminate() may be incorrect, or at the very least misleading.

`terminate()` currently explicitly returns ERROR, but if you ask it just after with api.get_status(), it will depend on what state it was when it was terminated. If it was running, it will be EXECUTED. If it was any other state I think the state will be unchanged.

I suspect terminate() returning ERROR is incorrect, as its not an error really. But it's never really been an issue before, as we generally follow a terminate with cleanup() so its all going way.

Possibly, terminate() should return `self.get_status(job)`, if it's going to be 100% accurate? The only reason to return ERROR would be if the docker kill failed, and that's handled by exceptions anyway. Not sure how much it matters, but would avoid the misleading ERROR. 

It may be we want to reshape some of the terminate/cleanup API as part of this work to ensure finalization anyway.